# Day 21: Step Counter
You manage to catch the airship right as it's dropping someone else off on their all-expenses-paid trip to Desert Island! It even helpfully drops you off near the gardener and his massive farm.

"You got the sand flowing again! Great work! Now we just need to wait until we have enough sand to filter the water for Snow Island and we'll have snow again in no time."

While you wait, one of the Elves that works with the gardener heard how good you are at solving problems and would like your help. He needs to get his steps in for the day, and so he'd like to know which garden plots he can reach with exactly his remaining 64 steps.

He gives you an up-to-date map (your puzzle input) of his starting position (S), garden plots (.), and rocks (#). For example:
```
...........
.....###.#.
.###.##..#.
..#.#...#..
....#.#....
.##..S####.
.##..#...#.
.......##..
.##.#.####.
.##..##.##.
...........
```
The Elf starts at the starting position (S) which also counts as a garden plot. Then, he can take one step north, south, east, or west, but only onto tiles that are garden plots. This would allow him to reach any of the tiles marked O:
```
...........
.....###.#.
.###.##..#.
..#.#...#..
....#O#....
.##.OS####.
.##..#...#.
.......##..
.##.#.####.
.##..##.##.
...........
```
Then, he takes a second step. Since at this point he could be at either tile marked O, his second step would allow him to reach any garden plot that is one step north, south, east, or west of any tile that he could have reached after the first step:
```
...........
.....###.#.
.###.##..#.
..#.#O..#..
....#.#....
.##O.O####.
.##.O#...#.
.......##..
.##.#.####.
.##..##.##.
...........
```
After two steps, he could be at any of the tiles marked O above, including the starting position (either by going north-then-south or by going west-then-east).

A single third step leads to even more possibilities:
```
...........
.....###.#.
.###.##..#.
..#.#.O.#..
...O#O#....
.##.OS####.
.##O.#...#.
....O..##..
.##.#.####.
.##..##.##.
...........
```
He will continue like this until his steps for the day have been exhausted. After a total of 6 steps, he could reach any of the garden plots marked O:
```
...........
.....###.#.
.###.##.O#.
.O#O#O.O#..
O.O.#.#.O..
.##O.O####.
.##.O#O..#.
.O.O.O.##..
.##.#.####.
.##O.##.##.
...........
```
In this example, if the Elf's goal was to get exactly 6 more steps today, he could use them to reach any of 16 garden plots.

However, the Elf actually needs to get 64 steps today, and the map he's handed you is much larger than the example map.

Starting from the garden plot marked S on your map, how many garden plots could the Elf reach in exactly 64 steps?

In [67]:
def result(inputfile, steps, start = None, startcolour = "w"):

    input = []

    with open(inputfile) as f:
        for line in f:
            input.append([*line.strip()])
    length = len(input)
    width = len(input[0])
    if start == None:
        for i, line in enumerate(input):
            if 'S' in line:
                start = [i, line.index('S')]
                break

    def neighbours(tile):
        x,y = tile
        neighbours = [[x + 1,y],[x - 1,y],[x,y + 1],[x,y- 1]]
        return [n for n in neighbours if 0 <= n[0] < length and 0 <= n[1] < width]

    whites = [start]
    blacks = []
    recentWhites = [start]
    recentBlacks = []
    for i in range(steps):
        if i % 2 == 0:
            for tile in recentWhites:
                for neighbour in neighbours(tile):
                    if input[neighbour[0]][neighbour[1]] != '#' and neighbour not in blacks:
                        blacks.append(neighbour)
                        recentBlacks.append(neighbour)
            recentWhites = []
        else:
            for tile in recentBlacks:
                for neighbour in neighbours(tile):
                    if input[neighbour[0]][neighbour[1]] != '#' and neighbour not in whites:
                        whites.append(neighbour)
                        recentWhites.append(neighbour)
            recentBlacks = []
    if (steps % 2 == 0) == (startcolour == "w"):
        return len(whites)
    else:
        return len(blacks)

In [4]:
result("example-input.txt", 6)

16

In [48]:
result("input.txt", 64)

3615

### Part Two
The Elf seems confused by your answer until he realizes his mistake: he was reading from a list of his favorite numbers that are both perfect squares and perfect cubes, not his step counter.

The actual number of steps he needs to get today is exactly 26501365.

He also points out that the garden plots and rocks are set up so that the map repeats infinitely in every direction.

So, if you were to look one additional map-width or map-height out from the edge of the example map above, you would find that it keeps repeating:
```
.................................
.....###.#......###.#......###.#.
.###.##..#..###.##..#..###.##..#.
..#.#...#....#.#...#....#.#...#..
....#.#........#.#........#.#....
.##...####..##...####..##...####.
.##..#...#..##..#...#..##..#...#.
.......##.........##.........##..
.##.#.####..##.#.####..##.#.####.
.##..##.##..##..##.##..##..##.##.
.................................
.................................
.....###.#......###.#......###.#.
.###.##..#..###.##..#..###.##..#.
..#.#...#....#.#...#....#.#...#..
....#.#........#.#........#.#....
.##...####..##..S####..##...####.
.##..#...#..##..#...#..##..#...#.
.......##.........##.........##..
.##.#.####..##.#.####..##.#.####.
.##..##.##..##..##.##..##..##.##.
.................................
.................................
.....###.#......###.#......###.#.
.###.##..#..###.##..#..###.##..#.
..#.#...#....#.#...#....#.#...#..
....#.#........#.#........#.#....
.##...####..##...####..##...####.
.##..#...#..##..#...#..##..#...#.
.......##.........##.........##..
.##.#.####..##.#.####..##.#.####.
.##..##.##..##..##.##..##..##.##.
.................................
```
This is just a tiny three-map-by-three-map slice of the inexplicably-infinite farm layout; garden plots and rocks repeat as far as you can see. The Elf still starts on the one middle tile marked S, though - every other repeated S is replaced with a normal garden plot (.).

Here are the number of reachable garden plots in this new infinite version of the example map for different numbers of steps:

- In exactly 6 steps, he can still reach 16 garden plots.
- In exactly 10 steps, he can reach any of 50 garden plots.
- In exactly 50 steps, he can reach 1594 garden plots.
- In exactly 100 steps, he can reach 6536 garden plots.
- In exactly 500 steps, he can reach 167004 garden plots.
- In exactly 1000 steps, he can reach 668697 garden plots.
- In exactly 5000 steps, he can reach 16733044 garden plots.

However, the step count the Elf needs is much larger! Starting from the garden plot marked S on your infinite map, how many garden plots could the Elf reach in exactly 26501365 steps?

In [78]:
def analyze(inputfile):

    input = []
    hashes = 0
    with open(inputfile) as f:
        for line in f:
            input.append([*line.strip()])
            hashes += line.count('#')
    length = len(input)
    width = len(input[0])
    for i, line in enumerate(input):
        if 'S' in line:
            s = [i, line.index('S')]
            break
    
    print("S-row: " + str(''.join(input[s[0]])))
    print("S-column: " + str(''.join([line[s[1]] for line in input])))

    print("length: " + str(length))
    print("width: " + str(width))
    print("S: " + str(s))
    print("steps left after reaching last S: " + str(26501365 % length))

    # get Time until full
    print("amount stones: " + str(hashes))
    amountTiles = length*width - hashes
    print("amount no stones: " + str(amountTiles))

    def neighbours(tile):
        x,y = tile
        neighbours = [[x + 1,y],[x - 1,y],[x,y + 1],[x,y- 1]]
        return [n for n in neighbours if 0 <= n[0] < length and 0 <= n[1] < width]

    whites = [s]
    blacks = []
    recentWhites = [s]
    recentBlacks = []
    i = 0
    noNewWhites = False
    noNewBlacks = False
    while(not(noNewWhites and noNewBlacks)):
        if i % 2 == 0:
            for tile in recentWhites:
                for neighbour in neighbours(tile):
                    if input[neighbour[0]][neighbour[1]] != '#' and neighbour not in blacks:
                        blacks.append(neighbour)
                        recentBlacks.append(neighbour)
            if len(recentWhites) == 0:
                noNewWhites = True
            recentWhites = []
        else:
            for tile in recentBlacks:
                for neighbour in neighbours(tile):
                    if input[neighbour[0]][neighbour[1]] != '#' and neighbour not in whites:
                        whites.append(neighbour)
                        recentWhites.append(neighbour)
            if len(recentBlacks) == 0:
                noNewBlacks = True
            recentBlacks = []
        i += 1
    print("time until fully explored: " + str(i))
    print("amount rechable white tiles: " + str(len(whites)))
    print("amount rechable black tiles: " + str(len(blacks)))
    print("amount rechable tiles: " + str(len(whites) + len(blacks)))

In [79]:
analyze("input.txt")

S-row: .................................................................S.................................................................
S-column: .................................................................S.................................................................
length: 131
width: 131
S: [65, 65]
steps left after reaching last S: 65
amount stones: 2422
amount no stones: 14739
time until fully explored: 133
amount rechable white tiles: 7354
amount rechable black tiles: 7362
amount rechable tiles: 14716


This means we get to the endpoint of a copy in each direction

In [27]:
26501365 / 131

202300.4961832061

We can traverse through 202300 copies in every direction.

Each copy has a different polarity then the neibouring ones.

The original returns its black tiles after a uneven number of steps.

The neighbouring 4 copies return their white tiles after the uneven number of steps.

The 8 neighbours of them "are" black again.

Their 12 neigbours are white agian.

In each new layer of copies we have 4*i new copies. We can go though 202299 layers until we reach the final layer that has to be treated specially because it wont be fully explored.

The other ones are all fully explored because you have at least 131 + 65 = 196 steps to explore them and 196 >= 133 which is the time until we are fully explored.

In [80]:
amountWhiteLayers = 0
amountBlackLayers = 1 #original

for i in range(1,202300):
    if i%2 == 0:
        amountBlackLayers += 4*i
    else:
        amountWhiteLayers += 4*i
print("amount white layers: " + str(amountWhiteLayers))
print("amount black layers: " + str(amountBlackLayers))
print("solution except final layer: " + str(amountWhiteLayers * 7354 + amountBlackLayers * 7362))

amount white layers: 40925290000
amount black layers: 40924885401
solution except final layer: 602253588982162


In [82]:
a1 = result("input.txt",130,[0,65])
a2 = result("input.txt",130,[65,0])
a3 = result("input.txt",130,[130,65])
a4 = result("input.txt",130,[65,130])
b1 = result("input.txt",64,[0,0])
b2 = result("input.txt",64,[0,130])
b3 = result("input.txt",64,[130,130])
b4 = result("input.txt",64,[130,0])
c1 = result("input.txt",195,[0,0])
c2 = result("input.txt",195,[0,130])
c3 = result("input.txt",195,[130,130])
c4 = result("input.txt",195,[130,0])
a = a1 + a2 + a3 + a4
b = b1 + b2 + b3 + b4
c = c1 + c2 + c3 + c4
print(a, b, c)

22192 3739 25820


We reach the `a` copies exactly 1 time, the `b` copies 202300 times and the `c` copies 202299 times.

![title](copy_explanation.png)

In [83]:
602253588982162 + a + 202300 * b + 202299 * c

602259568764234